File: counter/counter.py
#!/usr/bin/env python3
import argparse
import fcntl
import io
import json
import os
import re
import sys
from pathlib import Path
from urllib.parse import parse_qs

from PIL import Image, ImageDraw


ROOT = Path(__file__).resolve().parent
DATA_DIR = ROOT / "data"
DEFAULT_STRIP = ROOT / "counter-strip.gif"
DEFAULT_TOKENS = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", ","]
def main() -> None:
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("--query", default=None)
    parser.add_argument("--output", default=None)
    parser.add_argument("--no-headers", action="store_true")
    args = parser.parse_args()

    query_string = args.query if args.query is not None else os.environ.get("QUERY_STRING", "")
    payload = build_counter_gif(query_string)

    if args.output:
        Path(args.output).write_bytes(payload)
        return

    stdout = sys.stdout.buffer
    if not args.no_headers:
        stdout.write(b"Content-Type: image/gif\r\n")
        stdout.write(b"Cache-Control: no-store, no-cache, must-revalidate, max-age=0\r\n")
        stdout.write(b"Pragma: no-cache\r\n\r\n")
    stdout.write(payload)


def build_counter_gif(query_string: str) -> bytes:
    query = parse_qs(query_string, keep_blank_values=True)
    legacy = parse_legacy_options(query)
    strip_path = resolve_strip_path(legacy.get("dd", value(query, "dd", "")))
    counter_file = resolve_counter_file(
        value(query, "key", legacy.get("df", value(query, "df", "default")))
    )
    digits = clamp_int(value(query, "digits", "4"), 1, 12, 4)
    increment = value(query, "increment", "1") != "0"
    step = clamp_int(value(query, "step", "1"), 0, 1000, 1)
    use_comma = parse_legacy_bool(legacy.get("comma", value(query, "comma", "1")), True)
    frame_color = normalize_optional_color(legacy.get("frgb", value(query, "frgb", "")))
    frame_thickness = clamp_int(legacy.get("ft", value(query, "ft", "0")), 0, 6, 0)
    text_override = value(query, "text", "")

    DATA_DIR.mkdir(parents=True, exist_ok=True)
    count = read_and_update_count(counter_file, step if increment else 0)
    layout = load_strip_layout(strip_path)
    text = format_counter_text(count, digits, use_comma)
    tokens = tokenize_display_text(text_override, layout["tokens"]) if text_override else tokenize_display_text(text, layout["tokens"])
    image = render_counter_image(tokens, strip_path, layout, frame_color, frame_thickness)

    output = io.BytesIO()
    image.save(output, format="GIF")
    return output.getvalue()


def value(query: dict, key: str, default: str) -> str:
    values = query.get(key)
    if not values:
        return default
    return values[-1]


def clamp_int(raw: str, minimum: int, maximum: int, fallback: int) -> int:
    try:
        number = int(str(raw).strip())
    except ValueError:
        number = fallback
    return max(minimum, min(maximum, number))


def normalize_counter_key(raw: str) -> str:
    key = (raw or "").strip()
    key = re.sub(r"\.dat$", "", key, flags=re.IGNORECASE)
    key = re.sub(r"[^a-z0-9_.:()\-]+", "-", key, flags=re.IGNORECASE)
    key = key.strip("-").lower()
    return key[:120] if key else "default"


def resolve_counter_file(raw: str) -> Path:
    key = normalize_counter_key(raw)
    if key in {"", "default"}:
        key = "default"
    return DATA_DIR / f"{key}.dat"


def read_and_update_count(path: Path, increment_by: int) -> int:
    if not path.exists():
        path.write_text("0\n", encoding="utf-8")

    with path.open("r+", encoding="utf-8") as handle:
        fcntl.flock(handle.fileno(), fcntl.LOCK_EX)
        contents = handle.read().strip()
        count = max(0, int(contents or "0"))
        if increment_by > 0:
            count += increment_by
            handle.seek(0)
            handle.truncate()
            handle.write(f"{count}\n")
        fcntl.flock(handle.fileno(), fcntl.LOCK_UN)
    return count


def format_counter_text(count: int, minimum_digits: int, use_comma: bool) -> str:
    padded = str(max(0, count)).rjust(minimum_digits, "0")
    if not use_comma:
        return padded
    parts = []
    while len(padded) > 3:
        parts.insert(0, padded[-3:])
        padded = padded[:-3]
    parts.insert(0, padded)
    return ",".join(parts)


def parse_legacy_options(query: dict) -> dict:
    options = {}
    dd = value(query, "dd", "")
    df = value(query, "df", "")
    if df:
        options["df"] = df
    if not dd:
        return options
    parts = dd.split("|")
    options["dd"] = parts.pop(0)
    for part in parts:
        if "=" not in part:
            continue
        key, option_value = part.split("=", 1)
        key = key.strip().lower()
        option_value = option_value.strip()
        if key:
            options[key] = option_value
    return options


def parse_legacy_bool(raw: str, default: bool) -> bool:
    text = str(raw or "").strip().upper()
    if not text:
        return default
    if text in {"T", "TRUE", "1", "Y"}:
        return True
    if text in {"F", "FALSE", "0", "N"}:
        return False
    return default


def normalize_optional_color(raw: str):
    text = str(raw or "").strip().replace(";", "")
    if not text:
        return None
    if re.fullmatch(r"[0-9a-fA-F]{6}", text):
        return f"#{text.lower()}"
    if re.fullmatch(r"#[0-9a-fA-F]{6}", text):
        return text.lower()
    return None


def resolve_strip_path(style: str) -> Path:
    normalized = re.sub(r"[^@a-z0-9()_\-]+", "", style.strip().lower())
    candidates = []
    if normalized:
        candidates.append(ROOT / f"{normalized}.gif")
        candidates.append(ROOT / f"{normalized.replace('(', '').replace(')', '')}.gif")
    for candidate in candidates:
        if candidate.is_file():
            return candidate
    return DEFAULT_STRIP

def load_strip_layout(strip_path: Path) -> dict:
    metadata_path = strip_path.with_suffix(".meta.json")

    if metadata_path.is_file():
        data = json.loads(metadata_path.read_text(encoding="utf-8"))
        layout = parse_strip_metadata(data)
        if layout is not None:
            return layout

    with Image.open(strip_path) as strip:
        glyph_width = strip.width // len(DEFAULT_TOKENS)

    widths = [glyph_width] * len(DEFAULT_TOKENS)
    offsets = [index * glyph_width for index in range(len(DEFAULT_TOKENS))]
    return {"tokens": DEFAULT_TOKENS[:], "widths": widths, "offsets": offsets}


def parse_strip_metadata(data) -> dict | None:
    if not isinstance(data, dict):
        return None

    tokens = list(data.get("tokens", []))
    widths = [int(value) for value in data.get("widths", [])]
    offsets = [int(value) for value in data.get("offsets", [])]
    if tokens and len(tokens) == len(widths):
        if len(offsets) != len(widths):
            offsets = []
            left = 0
            for width in widths:
                offsets.append(left)
                left += width
        return {"tokens": tokens, "widths": widths, "offsets": offsets}

    base = data.get("base")
    if not isinstance(base, dict):
        return None

    tokens = list(base.get("tokens", []))
    if not tokens:
        return None

    base_width = int(base.get("width", 0) or 0)
    if base_width <= 0:
        return None

    widths = []
    offsets = []
    left = int(base.get("offset", 0) or 0)
    base_advance = int(base.get("advance", base_width) or base_width)

    for token in tokens:
        entry = data.get(token)
        width = base_width
        offset = left
        advance = base_advance

        if isinstance(entry, int):
            width = int(entry)
            advance = width
        elif isinstance(entry, dict):
            width = int(entry.get("width", base_width) or base_width)
            offset = int(entry.get("offset", left) or left)
            advance = int(entry.get("advance", width) or width)

        widths.append(width)
        offsets.append(offset)
        left = offset + advance

    return {"tokens": tokens, "widths": widths, "offsets": offsets}


def tokenize_display_text(raw_text: str, available_tokens) -> list[str]:
    tokens = []
    text = str(raw_text or "")
    index = 0
    token_set = set(available_tokens)

    while index < len(text):
        pair = text[index:index + 2].lower()
        if pair in {"am", "pm"} and pair in token_set:
            tokens.append(pair)
            index += 2
            continue

        character = text[index]
        if character in token_set:
            tokens.append(character)
        elif character.isdigit() and character in token_set:
            tokens.append(character)
        index += 1

    if not tokens:
        fallback = "0" if "0" in token_set else available_tokens[0]
        tokens.append(fallback)

    return tokens


def render_counter_image(tokens: list[str], strip_path: Path, layout: dict, frame_color, frame_thickness: int) -> Image.Image:
    strip = Image.open(strip_path).convert("P")
    glyph_height = strip.height
    token_to_index = {token: index for index, token in enumerate(layout["tokens"])}
    image_width = sum(layout["widths"][token_to_index.get(token, 0)] for token in tokens)
    image = Image.new("P", (image_width, glyph_height))
    image.putpalette(strip.getpalette())

    target_left = 0
    default_index = token_to_index.get("0", 0)

    for token in tokens:
        glyph_index = token_to_index.get(token, default_index)
        source_left = layout["offsets"][glyph_index]
        glyph_width = layout["widths"][glyph_index]
        cell = strip.crop((source_left, 0, source_left + glyph_width, glyph_height))
        image.paste(cell, (target_left, 0))
        target_left += glyph_width

    if frame_thickness > 0:
        image = apply_frame_thickness(image, frame_color or "#666666", frame_thickness)

    return image


def apply_frame_thickness(image: Image.Image, hex_color: str, thickness: int) -> Image.Image:
    framed = Image.new("RGBA", (image.width + (thickness * 2), image.height + (thickness * 2)), (255, 0, 255, 0))
    framed.alpha_composite(image.convert("RGBA"), (thickness, thickness))
    rgb_image = framed.convert("RGB")
    draw = ImageDraw.Draw(rgb_image)
    outline = hex_to_rgb(hex_color)
    top = 0
    left = 0
    right = rgb_image.width - 1
    bottom = rgb_image.height - 1

    for line in range(thickness):
        draw.rectangle((left + line, top + line, right - line, bottom - line), outline=outline)

    return rgb_image.convert("P", palette=Image.ADAPTIVE, colors=256)


def hex_to_rgb(hex_color: str):
    value = hex_color.lstrip("#")
    return tuple(int(value[index:index + 2], 16) for index in (0, 2, 4))


if __name__ == "__main__":
    main()

File: counter/counter.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"image"
	"image/color"
	"image/draw"
	"image/gif"
	"io"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
)

type stripLayout struct {
	Tokens  []string `json:"tokens"`
	Widths  []int    `json:"widths"`
	Offsets []int
}

type compactStripBase struct {
	Tokens  []string `json:"tokens"`
	Width   int      `json:"width"`
	Offset  int      `json:"offset"`
	Advance int      `json:"advance"`
}

type compactStripEntry struct {
	Width   int `json:"width"`
	Offset  int `json:"offset"`
	Advance int `json:"advance"`
}

var (
	rootPath, _     = filepath.Abs(filepath.Dir(os.Args[0]))
	dataDir         = filepath.Join(rootPath, "data")
	defaultStrip    = filepath.Join(rootPath, "counter-strip.gif")
	defaultTokens   = []string{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", ","}
	protectedColors = map[[3]uint8]bool{
		{0, 0, 0}:       true,
		{232, 232, 232}: true,
		{252, 252, 252}: true,
		{168, 168, 168}: true,
	}
	keySanitizer   = regexp.MustCompile(`[^a-z0-9_.:()\-]+`)
	stripSanitizer = regexp.MustCompile(`[^@a-z0-9()_\-]+`)
)

func main() {
	queryArg := flag.String("query", "", "")
	outputArg := flag.String("output", "", "")
	noHeaders := flag.Bool("no-headers", false, "")
	flag.Parse()

	queryString := *queryArg
	if queryString == "" {
		queryString = os.Getenv("QUERY_STRING")
	}

	payload, err := buildCounterGIF(queryString)
	if err != nil {
		if *outputArg != "" {
			_ = os.WriteFile(*outputArg, []byte(err.Error()), 0o644)
			os.Exit(1)
		}
		io.WriteString(os.Stdout, "Status: 500 Internal Server Error\r\n")
		io.WriteString(os.Stdout, "Content-Type: text/plain; charset=utf-8\r\n\r\n")
		io.WriteString(os.Stdout, err.Error())
		os.Exit(1)
	}

	if *outputArg != "" {
		_ = os.WriteFile(*outputArg, payload, 0o644)
		return
	}

	if !*noHeaders {
		io.WriteString(os.Stdout, "Content-Type: image/gif\r\n")
		io.WriteString(os.Stdout, "Cache-Control: no-store, no-cache, must-revalidate, max-age=0\r\n")
		io.WriteString(os.Stdout, "Pragma: no-cache\r\n\r\n")
	}
	_, _ = os.Stdout.Write(payload)
}

func buildCounterGIF(queryString string) ([]byte, error) {
	values, _ := url.ParseQuery(queryString)
	legacy := parseLegacyOptions(values)
	stripPath := resolveStripPath(firstNonEmpty(values.Get("strip"), legacy["dd"], values.Get("dd")))
	counterFile := resolveCounterFile(firstNonEmpty(values.Get("key"), legacy["df"], values.Get("df"), "default"))
	digits := clampInt(values.Get("digits"), 1, 12, 4)
	increment := firstNonEmpty(values.Get("increment"), "1") != "0"
	step := clampInt(values.Get("step"), 0, 1000, 1)
	useComma := parseLegacyBool(firstNonEmpty(legacy["comma"], values.Get("comma"), "1"), true)
	frameColor := normalizeOptionalColor(firstNonEmpty(legacy["frgb"], values.Get("frgb")))
	frameThickness := clampInt(firstNonEmpty(legacy["ft"], values.Get("ft"), "0"), 0, 6, 0)
	textOverride := values.Get("text")

	_ = os.MkdirAll(dataDir, 0o755)
	count, err := readAndUpdateCount(counterFile, ternaryInt(increment, step, 0))
	if err != nil {
		return nil, err
	}

	layout, err := loadStripLayout(stripPath)
	if err != nil {
		return nil, err
	}

	text := formatCounterText(count, digits, useComma)
	var tokens []string
	if textOverride != "" {
		tokens = tokenizeDisplayText(textOverride, layout.Tokens)
	} else {
		tokens = tokenizeDisplayText(text, layout.Tokens)
	}

	imageResult, err := renderCounterImage(tokens, stripPath, layout, frameColor, frameThickness)
	if err != nil {
		return nil, err
	}

	var buffer bytes.Buffer
	if err := gif.Encode(&buffer, imageResult, nil); err != nil {
		return nil, err
	}
	return buffer.Bytes(), nil
}

func parseLegacyOptions(values url.Values) map[string]string {
	options := map[string]string{}
	if df := values.Get("df"); df != "" {
		options["df"] = df
	}
	dd := values.Get("dd")
	if dd == "" {
		return options
	}
	parts := strings.Split(dd, "|")
	options["dd"] = parts[0]
	for _, part := range parts[1:] {
		pair := strings.SplitN(part, "=", 2)
		if len(pair) != 2 {
			continue
		}
		key := strings.ToLower(strings.TrimSpace(pair[0]))
		value := strings.TrimSpace(pair[1])
		if key != "" {
			options[key] = value
		}
	}
	return options
}

func normalizeCounterKey(raw string) string {
	key := strings.TrimSpace(raw)
	key = strings.TrimSuffix(strings.ToLower(key), ".dat")
	key = keySanitizer.ReplaceAllString(key, "-")
	key = strings.Trim(key, "-")
	if key == "" {
		return "default"
	}
	if len(key) > 120 {
		return key[:120]
	}
	return key
}

func resolveCounterFile(raw string) string {
	key := normalizeCounterKey(raw)
	if key == "" {
		key = "default"
	}
	return filepath.Join(dataDir, key+".dat")
}

func readAndUpdateCount(counterFile string, incrementBy int) (int, error) {
	if _, err := os.Stat(counterFile); os.IsNotExist(err) {
		if err := os.WriteFile(counterFile, []byte("0\n"), 0o644); err != nil {
			return 0, err
		}
	}

	file, err := os.OpenFile(counterFile, os.O_RDWR, 0o644)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	if err := syscall.Flock(int(file.Fd()), syscall.LOCK_EX); err != nil {
		return 0, err
	}
	defer syscall.Flock(int(file.Fd()), syscall.LOCK_UN)

	contents, err := io.ReadAll(file)
	if err != nil {
		return 0, err
	}
	count, _ := strconv.Atoi(strings.TrimSpace(string(contents)))
	if count < 0 {
		count = 0
	}

	if incrementBy > 0 {
		count += incrementBy
		if _, err := file.Seek(0, 0); err != nil {
			return 0, err
		}
		if err := file.Truncate(0); err != nil {
			return 0, err
		}
		if _, err := file.WriteString(strconv.Itoa(count) + "\n"); err != nil {
			return 0, err
		}
	}

	return count, nil
}

func clampInt(raw string, minValue int, maxValue int, fallback int) int {
	number, err := strconv.Atoi(strings.TrimSpace(raw))
	if err != nil {
		number = fallback
	}
	if number < minValue {
		return minValue
	}
	if number > maxValue {
		return maxValue
	}
	return number
}

func parseLegacyBool(raw string, fallback bool) bool {
	text := strings.ToUpper(strings.TrimSpace(raw))
	switch text {
	case "T", "TRUE", "1", "Y":
		return true
	case "F", "FALSE", "0", "N":
		return false
	case "":
		return fallback
	default:
		return fallback
	}
}

func normalizeOptionalColor(raw string) *[3]uint8 {
	text := strings.ReplaceAll(strings.TrimSpace(raw), ";", "")
	if text == "" {
		return nil
	}
	if strings.HasPrefix(text, "#") {
		text = text[1:]
	}
	if len(text) != 6 {
		return nil
	}
	value, err := strconv.ParseUint(text, 16, 32)
	if err != nil {
		return nil
	}
	result := [3]uint8{uint8(value >> 16), uint8((value >> 8) & 0xFF), uint8(value & 0xFF)}
	return &result
}

func resolveStripPath(style string) string {
	normalized := stripSanitizer.ReplaceAllString(strings.ToLower(strings.TrimSpace(style)), "")
	candidates := []string{}
	if normalized != "" {
		if strings.HasSuffix(normalized, ".gif") {
			candidates = append(candidates, filepath.Join(rootPath, normalized))
		} else {
			candidates = append(candidates, filepath.Join(rootPath, normalized+".gif"))
		}
		candidates = append(candidates, filepath.Join(rootPath, strings.ReplaceAll(strings.ReplaceAll(normalized, "(", ""), ")", "")+".gif"))
	}
	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
	}
	return defaultStrip
}

func loadStripLayout(stripPath string) (stripLayout, error) {
	metadataPath := strings.TrimSuffix(stripPath, filepath.Ext(stripPath)) + ".meta.json"
	if contents, err := os.ReadFile(metadataPath); err == nil {
		if layout, ok := parseStripMetadata(contents); ok {
			return layout, nil
		}
	}

	file, err := os.Open(stripPath)
	if err != nil {
		return stripLayout{}, err
	}
	defer file.Close()

	source, err := gif.Decode(file)
	if err != nil {
		return stripLayout{}, err
	}
	glyphWidth := source.Bounds().Dx() / len(defaultTokens)
	widths := make([]int, len(defaultTokens))
	offsets := make([]int, len(defaultTokens))
	for index := range defaultTokens {
		widths[index] = glyphWidth
		offsets[index] = index * glyphWidth
	}
	return stripLayout{Tokens: append([]string{}, defaultTokens...), Widths: widths, Offsets: offsets}, nil
}

func parseStripMetadata(contents []byte) (stripLayout, bool) {
	var layout stripLayout
	if err := json.Unmarshal(contents, &layout); err == nil && len(layout.Tokens) == len(layout.Widths) && len(layout.Tokens) > 0 {
		if len(layout.Offsets) != len(layout.Widths) {
			layout.Offsets = layout.Offsets[:0]
			left := 0
			for _, width := range layout.Widths {
				layout.Offsets = append(layout.Offsets, left)
				left += width
			}
		}
		return layout, true
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(contents, &raw); err != nil {
		return stripLayout{}, false
	}

	var base compactStripBase
	if err := json.Unmarshal(raw["base"], &base); err != nil || len(base.Tokens) == 0 || base.Width <= 0 {
		return stripLayout{}, false
	}

	baseAdvance := base.Advance
	if baseAdvance <= 0 {
		baseAdvance = base.Width
	}

	result := stripLayout{
		Tokens: append([]string{}, base.Tokens...),
		Widths: make([]int, 0, len(base.Tokens)),
		Offsets: make([]int, 0, len(base.Tokens)),
	}

	left := base.Offset
	for _, token := range base.Tokens {
		width := base.Width
		offset := left
		advance := baseAdvance

		if entryRaw, ok := raw[token]; ok {
			var widthOnly int
			if err := json.Unmarshal(entryRaw, &widthOnly); err == nil {
				width = widthOnly
				advance = width
			} else {
				var entry compactStripEntry
				if err := json.Unmarshal(entryRaw, &entry); err == nil {
					if entry.Width > 0 {
						width = entry.Width
					}
					if entry.Offset != 0 || left == 0 {
						offset = entry.Offset
					}
					if entry.Advance > 0 {
						advance = entry.Advance
					} else {
						advance = width
					}
				}
			}
		}

		result.Widths = append(result.Widths, width)
		result.Offsets = append(result.Offsets, offset)
		left = offset + advance
	}

	return result, true
}

func tokenizeDisplayText(raw string, available []string) []string {
	tokenSet := map[string]bool{}
	for _, token := range available {
		tokenSet[token] = true
	}

	tokens := []string{}
	for index := 0; index < len(raw); {
		if index+2 <= len(raw) {
			pair := strings.ToLower(raw[index : index+2])
			if (pair == "am" || pair == "pm") && tokenSet[pair] {
				tokens = append(tokens, pair)
				index += 2
				continue
			}
		}
		character := raw[index : index+1]
		if tokenSet[character] {
			tokens = append(tokens, character)
		}
		index++
	}

	if len(tokens) == 0 {
		if tokenSet["0"] {
			return []string{"0"}
		}
		return []string{available[0]}
	}

	return tokens
}

func formatCounterText(count int, minimumDigits int, useComma bool) string {
	if count < 0 {
		count = 0
	}
	padded := strconv.Itoa(count)
	for len(padded) < minimumDigits {
		padded = "0" + padded
	}
	if !useComma {
		return padded
	}
	parts := []string{}
	for len(padded) > 3 {
		parts = append([]string{padded[len(padded)-3:]}, parts...)
		padded = padded[:len(padded)-3]
	}
	parts = append([]string{padded}, parts...)
	return strings.Join(parts, ",")
}

func renderCounterImage(tokens []string, stripPath string, layout stripLayout, frameColor *[3]uint8, frameThickness int) (image.Image, error) {
	file, err := os.Open(stripPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	source, err := gif.Decode(file)
	if err != nil {
		return nil, err
	}

	sourceRGBA := toRGBA(source)
	glyphHeight := sourceRGBA.Bounds().Dy()
	tokenIndex := map[string]int{}
	for index, token := range layout.Tokens {
		tokenIndex[token] = index
	}
	defaultIndex := tokenIndex["0"]

	imageWidth := 0
	for _, token := range tokens {
		glyphIndex, ok := tokenIndex[token]
		if !ok {
			glyphIndex = defaultIndex
		}
		imageWidth += layout.Widths[glyphIndex]
	}

	target := image.NewRGBA(image.Rect(0, 0, imageWidth, glyphHeight))
	left := 0
	for _, token := range tokens {
		glyphIndex, ok := tokenIndex[token]
		if !ok {
			glyphIndex = defaultIndex
		}
		width := layout.Widths[glyphIndex]
		srcRect := image.Rect(layout.Offsets[glyphIndex], 0, layout.Offsets[glyphIndex]+width, glyphHeight)
		draw.Draw(target, image.Rect(left, 0, left+width, glyphHeight), sourceRGBA, srcRect.Min, draw.Src)
		left += width
	}

	if frameThickness > 0 {
		target = applyFrameThickness(target, *valueOrDefault(frameColor, [3]uint8{102, 102, 102}), frameThickness)
	}

	return target, nil
}

func toRGBA(img image.Image) *image.RGBA {
	bounds := img.Bounds()
	rgba := image.NewRGBA(bounds)
	draw.Draw(rgba, bounds, img, bounds.Min, draw.Src)
	return rgba
}

func applyFrameThickness(img *image.RGBA, frame [3]uint8, thickness int) *image.RGBA {
	frameColor := color.RGBA{R: frame[0], G: frame[1], B: frame[2], A: 255}
	bounds := img.Bounds()
	framed := image.NewRGBA(image.Rect(0, 0, bounds.Dx()+(thickness*2), bounds.Dy()+(thickness*2)))
	draw.Draw(framed, image.Rect(thickness, thickness, thickness+bounds.Dx(), thickness+bounds.Dy()), img, bounds.Min, draw.Src)

	for line := 0; line < thickness; line++ {
		rect := image.Rect(line, line, framed.Bounds().Dx()-line, framed.Bounds().Dy()-line)
		strokeRect(framed, rect, frameColor)
	}

	return framed
}

func strokeRect(img *image.RGBA, rect image.Rectangle, c color.RGBA) {
	maxX := rect.Max.X - 1
	maxY := rect.Max.Y - 1
	for x := rect.Min.X; x <= maxX; x++ {
		img.SetRGBA(x, rect.Min.Y, c)
		img.SetRGBA(x, maxY, c)
	}
	for y := rect.Min.Y; y <= maxY; y++ {
		img.SetRGBA(rect.Min.X, y, c)
		img.SetRGBA(maxX, y, c)
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}

func ternaryInt(condition bool, whenTrue int, whenFalse int) int {
	if condition {
		return whenTrue
	}
	return whenFalse
}

func valueOrDefault(value *[3]uint8, fallback [3]uint8) *[3]uint8 {
	if value != nil {
		return value
	}
	return &fallback
}
